src/Simulation/Simulators/SparseSimulator/SparseSimulatorCS/Probes.cs

```diff
@@ -0,0 +1,360 @@
+                // Returns false if the state was entangled.
+                if (!((SparseSimulatorProcessor)this.QuantumProcessor).Dump(callback, ids.Max(), qubits))
+                {
+                    channel("## Qubits were entangled with an external qubit. Cannot dump corresponding wave function. ##");
+                }
+
+                return QVoid.Instance;
+            }
+
+            // This generates the channel, then calls it
+            // If no file provided, use `Message` to generate the message into the console;
+            if (string.IsNullOrWhiteSpace(filename))
+            {
+                var op = this.Get<ICallable<string, QVoid>, Microsoft.Quantum.Intrinsic.Message>();
+                return process((msg) => op.Apply(msg));
+            }
+            else
+            {
+                try
+                {
+                    using (var file = new StreamWriter(filename))
+                    {
+                        return process(file.WriteLine);
+                    }
+                }
+                catch (Exception e)
+                {
+                    logMessage.Apply($"[warning] Unable to write state to '{filename}' ({e.Message})");
+                    return QVoid.Instance;
+                }
+            }
+        }
+
+
+        // Called by the diagnostics functions
+        public class SparseSimDumpMachine<T> : Quantum.Diagnostics.DumpMachine<T>
+        {
+            private SparseSimulator Simulator { get; }
+
+            public SparseSimDumpMachine(SparseSimulator m) : base(m)
+            {
+                this.Simulator = m;
+            }
+
+            public override Func<T, QVoid> __Body__ => (location) =>
+            {
+                if (location == null) { throw new ArgumentNullException(nameof(location)); }
+
+                return Simulator.Dump(location);
+            };
+        }
+
+        // Called by the diagnostics functions
+        public class QSimDumpRegister<T> : Quantum.Diagnostics.DumpRegister<T>
+        {
+            private SparseSimulator Simulator { get; }
+
+
+            public QSimDumpRegister(SparseSimulator m) : base(m)
+            {
+                this.Simulator = m;
+            }
+
+            public override Func<(T, IQArray<Qubit>), QVoid> __Body__ => (__in) =>
+            {
+                var (location, qubits) = __in;
+
+                if (location == null) { throw new ArgumentNullException(nameof(location)); }
+
+                return Simulator.Dump(location, qubits);
+            };
+        }
+
+        private string FormatMagnitude(double magnitude, double phase) =>
+                (new String('*', (int)System.Math.Ceiling(20.0 * magnitude))).PadRight(20) + $" [ {magnitude:F6} ]";
+
+        /// <summary>
+        /// Returns a string that represents the phase of the amplitude.
+        /// </summary>
+        private string FormatAngle(double magnitude, double angle)
```

Comment:
This looks to be duplicated with other angle-formatting logic; perhaps it would be good to isolate both into a new static class with utilities for such formatting?